Allow awaiting futures automatically in parameters

blastrock opened the pull request:
Async fixture are kind of hard to use currently as reported in #157 . This PR adds a syntax to make tests with async fixtures easier to write.

Before:

```rust
    #[rstest]
    async fn it_can_return_its_ip(#[future] client: Client) {
        let client = client.await;
        client.doThis();
        client.doThat();
    }
```

After:

```rust
    #[rstest]
    async fn it_can_return_its_ip(#[await_future] client: Client) {
        client.doThis();
        client.doThat();
    }
```

This PR is ready code-wise, but it still lacks some tests and documentation. I wanted to get feedback for this approach before going further, please tell me what you think :)

la10736 commented:
First of all thanks too much for your effort and ... WOW it's a lot of work!!!!

Please be patient.... I've not too much time to look into it. I hope to find some times in the next days (1 or 2 weeks ... sorry :()

blastrock commented:
No problem, I know what it is to manage an open source project, no pressure :)

Initially I just did it for fun and to discover how macro worked in Rust, then it become something more serious ^^

la10736 commented on rstest_macros/src/render/mod.rs:
```diff
@@ -260,6 +280,19 @@ fn single_test_case<'a>(
     };
     let execute = render_test_call(testfn_name.clone().into(), &args, timeout, is_async);
 
+    let test_impl = test_impl.map(|test_impl| {
+        let mut test_impl = test_impl.clone();
+        for arg in test_impl.sig.inputs.iter_mut() {
+            if let FnArg::Typed(t) = arg {
+                t.attrs = std::mem::take(&mut t.attrs)
+                    .into_iter()
+                    .filter(|a| !attr_is(a, "future"))
+                    .collect();
+            }
+        }
+        test_impl
+    });
+
     quote! {
```
Why you're removing the `#[future]` attrs?

blastrock replied:
Whoops, I think that's code that I left over from when I was hacking the solution, I can remove it now

la10736 commented on rstest_macros/src/render/mod.rs:
```diff
@@ -57,14 +59,18 @@ pub(crate) fn single(mut test: ItemFn, info: RsTestInfo) -> TokenStream {
+        await_args,
+    } = info;
     let resolver_fixtures = resolver::fixtures::get(data.fixtures());
 
     let rendered_cases = cases_data(&data, test.sig.ident.span())
         .map(|(name, attrs, resolver)| {
             TestCaseRender::new(name, attrs, (resolver, &resolver_fixtures))
         })
-        .map(|case| case.render(&test, &attributes))
+        .map(|case| case.render(&test, &await_args, &attributes))
         .collect();
```
I'm quite sure that we can avoid to spread `await_args` in all functions signature... I think that `test` and `attributes` can contains all information. I we start to add a new argument for each feature the code will become a mess (more than it already is :) ).

la10736 reviewed (commented):
Sorry, maybe I've lost the big plan... What's the real difference between `#[future]`and`#[future_awit]`? Just add the await call on fixture resolution? 
In this case I guess that the code cold be really simplified and I handle this case as a special case of the first one. (I'm writing this by mobile)

Anyway... follow my review of this code.

Great and huge work!!! But we should fix some things to this draft 
1. Rebase it to main
2. Avoid to mix render and parse: parse should add all information to enable render to write the correct code
3. Use `Resolver` composability  instead of add new field to `ArgumentResolver` struct
4. Add tests for catching error in parsing
5. Find a way to don't spread `async_args` in all function signature

Thanks again for your effort. I'll try to support your work by doing my best and give you a fast (I hope) feedback.

blastrock commented on rstest_macros/src/parse/future.rs:
```diff
@@ -100,6 +117,108 @@ impl VisitMut for ReplaceFutureAttribute {
+                    if tr.lifetime.is_none() {
+                        let lifetime = syn::Lifetime {
+                            apostrophe: ident.span(),
+                            ident: format_ident!("_{}", ident),
```
This code comes from the ReplaceFutureAttribute type just above. At the beginning I replaced it completely, then I thought that I shouldn't break the API. I ended up with a copy paste that maybe can be factorized...

But for this particular issue, I just took the code from above ^^

la10736 replied:
Ok... It should not break any API but `_` is enough

blastrock commented on rstest_macros/src/render/fixture.rs:
```diff
@@ -72,14 +75,33 @@ pub(crate) fn render(fixture: ItemFn, info: FixtureInfo) -> TokenStream {
         impl #name {
             #(#orig_attrs)*
             #[allow(unused_mut)]
-            pub #asyncness fn get #generics (#orig_args) #output #where_clause {
```
No particular issue. It's just that I changed this API, and when I saw the "pub" I thought "oh, maybe I can't change it after all, it'll affect a lot of code", but actually the function wasn't used anywhere so I just removed the pub. I thought it would be better to add pub if I need it than to put it by default.

blastrock commented:
Thanks for the review! I'm a bit busy too, I'll try to apply the changes this weekend.

> Sorry, maybe I've lost the big plan... What's the real difference between #[future]and#[future_awit]? Just add the await call on fixture resolution?

Yes, that's it, just the await call.

>  In this case I guess that the code cold be really simplified and I handle this case as a special case of the first one. (I'm writing this by mobile)

I can't remember exactly but I couldn't reuse the `future` code. The `future` attr is handled by just replacing the type `T` with `impl Future<T>` . In the case of await we need to generate the await code in the wrapper code of fixtures and test cases.

la10736 commented:
OOOHHHHH!!! You're right ... also the code that handle `future` is a mess :smile:

Ok, we can start to include your version and after refactor them. In parsing stage we should just identify the attributes,
remove them and accumulate the info about what arguments need `impl Future` modification and what need `await`.
On render stage use these information to modify type and resolver.

blastrock commented on rstest_macros/src/parse/fixture.rs:
```diff
@@ -21,10 +21,23 @@ use crate::{parse::Attribute, utils::attr_in};
 pub(crate) struct FixtureInfo {
     pub(crate) data: FixtureData,
     pub(crate) attributes: FixtureModifiers,
+    pub(crate) signature_with_future_impls: Signature,
```
Done.

I'll mark conversations as resolved as I go through them, I'll push only at the end when I have something worth showing :)

blastrock commented:
Ok, I managed to work something out! I need some feedback though.

As usual, no hurry :) When you have some time, I'd like you to check the commit 51d626c . This is something I needed to use a resolver to add the await expression. I moved the decision to use a fixture in the parsing stage instead of the end of the injection stage. There is one downside to this, it's that the `extract_fixtures` call must be first (I put a comment there so that you notice it), I know this is something you didn't like in my first patch but I couldn't find a better idea, tell me if you have one.

Then, you can just review the `AwaitResolver` part in the "--wip--" commit and tell me if it is the approach you were thinking of.

Also note that all integration tests pass, but I broke a lot of unit tests, so no need to review in too much detail, this is far from the final version. And I didn't apply all the changes you requested yet ^^

la10736 commented:
Ok, I spend all the morning to try to recover from my mind what I was doing in my code :smile:. I also started to write something new from scratch.... 
We need to introduce something new here: all resolvable variable should have an `Option` like 

```rust
#[derive(PartialEq, Debug, Clone)]
pub(crate) enum AsyncType {
    Plain(Type),
    Await(Type),
}
```
Ok, it's simple to add it to `Fixture` struct but we need it also in case and values that use plain `Ident` for now :cry:

We need an exit strategy: Forget all my comments about don't mix parser and render phases and also for resolver part. I'll merge your PR when you'll fix all comments and later I'll open a ticket for refactor all code and remove this mess.

I'm sorry .... we're paying an hack that I did to introduce `#[future]` attribute as fast as possible.

la10736 commented:
I spend all the morning to understand what I wrote and why :smile:

We should introduce a new attribute to all items that should be resolved: an `Option` of 

```rust
#[derive(PartialEq, Debug, Clone)]
pub(crate) enum AsyncType {
    Plain(Type),
    Await(Type),
}
```
with this information we can write the correct render and replace type if we need. 

But we need it not just for `Fixture` but also for values and case args that unfortunately are implemented as plain `Ident` :cry:

We need an exit strategy. So, forget all my comments about don't mix parse and render stage (also comments about resolver); fix all others notes and add the missing tests and I'll merge your PR. After I'll open a new ticket to handle the complete refactoring to clean and generalize the implementation.

I'm sorry, we're paying an hack that I've introduced to implement `#[future]` faster for all cases....

blastrock commented:
Here it is. I kept the last version with the resolvers, fixed all the tests and addressed your comments. I didn't write new unit tests since you're going to refactor and they'll probably be more pain to refactor.

>  We need to introduce something new here: all resolvable variable should have an Option like

That's an option, but it will change a lot of code, much like my first solution. I let you experiment with that and be the judge ^^

Tell me if you need some more changes before merging :)

Also, if I can add one more controversial opinion: I know I'm the one who chose it but `await_future` is kind of a long name. I initially wanted to just write `#[await]` but you can't put a keyword in there. Maybe we should go with an abbreviation `#[awt]`?

la10736 commented:
> Also, if I can add one more controversial opinion: I know I'm the one who chose it but await_future is kind of a long name. I initially wanted to just write #[await] but you can't put a keyword in there. Maybe we should go with an abbreviation #[awt]?

Also I don't like this name and I don't like that is _another kind of `#[future]`_ ... so what about `#[future(awt)]`? ... ok is longer but don't introduce a new _keyword_ but just an option on an existing one.

pleshevskiy commented:
I just started using your library for integration tests and I absolutely do not understand for what cases we need to always manually `await` for async fixtures? Can't we automatically `await` for them in the order we specified in test_case function?

pleshevskiy commented:
At worst, I would prefer a single modifier before the test_case function. And instead of writing `#[await_future]` / `#[future(awt)]` 

For example:
```rust
#[rstest]
#[await_futures]
async fn it_can_return_its_ip(
    #[future] di_container: RootModule,
    #[future] client: Client,
) {
    client.doThis();
    client.doThat();
}
```

la10736 commented:
There are a lot of cases where the code that you should test take a non _awaited_ future bu just a future: fixture can provide this kind of references and `#[future]` is something that help you on remove the boiler plate on receiving a `Future`. If you don't need a future from your fixture why don't create a concrete (await for the future) and return it like a standard fixture? 

Ok... `#[awt]` is good

pleshevskiy commented:
Or maybe we're approaching this problem from the wrong side. What if we add a macro for `fixture` to do a wrapper over the async function and create a convenient wrapper for us?

```
#[fixture]
#[block]
pub async fn di_container() -> DiContainer {
    let db_pool = PgConnectionPool::create(CreateDbConnectionPoolParams {
        host: "localhost",
        port: 6666,
        user: "test",
        password: "test",
        db_name: "test",
    })
    .await;

    create_di_container(db_pool)
}


#[fixture]
#[block]
pub async fn client(di_container: DiContainer) -> TestClient {
    let app = create_app(CreateAppParams {
        di_container: Arc::new(di_container.await),
    });

    TestClient::new(app)
}
```

pleshevskiy commented:
I think this approach would also solve issue https://github.com/la10736/rstest/issues/141

la10736 commented:
I close this PR in favor to #186 
Thank very much for your effort!!!!
